dll/win32/kernel32/kernel32.spec

```diff
@@ -482,7 +482,10 @@
 @ stub -version=0x600+ GetFinalPathNameByHandleA
 @ stub -version=0x600+ GetFinalPathNameByHandleW
 @ stdcall GetFirmwareEnvironmentVariableA(str str ptr long)
+@ stdcall -version=0x602+ GetFirmwareEnvironmentVariableExA(str str ptr long ptr)
+@ stdcall -version=0x602+ GetFirmwareEnvironmentVariableExW(wstr wstr ptr long ptr)
 @ stdcall GetFirmwareEnvironmentVariableW(wstr wstr ptr long)
+@ stdcall -version=0x602+ GetFirmwareType(ptr)
```

Comment:
Question for @tkreuzer / @learn-more (perhaps) : what's currently the best place to get the new _implemented_ (not stubbed) NT6+ functions? Here, with the version thing, or in kernel32_vista ?

Reply:
Well, I'll :
+ Move `GetFirmwareType` to kernel32_vista
+ **Copy** (Get/Set)FirmwareEnvironmentVariableEx[A/W] to kernel32_vista, because (Get/Set)FirmwareEnvironmentVariable call their Ex version internally, code will have redundance.

Is that okay? @HBelusca

Reply:
Seems hard to do so:

In `k32_vista.h`, `NTDDI_VERSION` is lower than `NTDDI_LONGHORN`:
```
/* Redefine NTDDI_VERSION to 2K3 SP1 to get correct NDK definitions */
#undef NTDDI_VERSION
#define NTDDI_VERSION NTDDI_WS03SP1
```
So kernel32_vista cannot has `SYSTEM_BOOT_ENVIRONMENT_INFORMATION` definition, that's `GetFirmwareType` required. Define it again in kernel32_vista seems is not a good programming practice.

And also, re-implement the "Ex" version in kernel32_vista, code will have a big redundance, I prefer to give up exporting them and keep them in kernel32 as private functions for now.

BTW, those UM API is a part of UEFI, that's NT6.2+ stuff. If we support UEFI, could we export them whatever NT >= 6.2 or not?

Reply:
If it were me, I'd create a new k32_vista (k32_win8 lol?) header that only is included in the file you make to store these APIs in. 
But maybe that's not a great solution either to be honest as that implies quite a bit of redundancy for something a single thing.

But to answer your other question here from my opinion;
I don't think ROS will have a habit of using these anytime soon. There's not really any app one should be running in ROS anyway that needs to modify the firmware variables. So I don't think it's worth exporting no matter the NT version. 

However in the not to distant road these would eventually need to be used if we want to install on a modern system the "Correct way". Cross that bridge when we get there I'd say.

Reply:
> 

I'm planned to remove NT6+ functions in this PR:
+ Remove `GetFirmwareType` and `_FIRMWARE_TYPE`
+ Do not export `(Get/Set)FirmwareEnvironmentVariableEx[A/W]` but keep as internal functions used by `(Get/Set)FirmwareEnvironmentVariable[A/W]`
+ Keep `_SYSTEM_BOOT_ENVIRONMENT_INFORMATION` and API tests for future

How about this plan? BTW, some commands to edit boot sequence, and "Boot to BIOS setup next time" feature will use them. For example, "shutdown /r **/fw**" is a real OS app call them to modify firmware variables, to change the boot sequence make machine enter BIOS setup next time. @DarkFire01